src/dataflow/src/render/upsert.rs

```diff
@@ -309,3 +320,125 @@ where
+                        position: new_position,
+                        upstream_time_millis: new_upstream_time_millis,
+                    },
+                    time,
+                ) in vector.drain(..)
+                {
+                    if key.is_empty() {
+                        error!("{}", "Encountered empty key");
+                        continue;
+                    }
+                    let entry = to_send
+                        .entry(time)
+                        .or_insert_with(|| (cap.delayed(&time), HashMap::new()))
+                        .1
+                        .entry(key)
+                        .or_insert_with(Default::default);
+
+                    let new_entry = SourceData {
+                        value: new_value,
+                        position: new_position,
+                        upstream_time_millis: new_upstream_time_millis,
+                    };
+
+                    // if the time is equal, toss out the row with the
+                    // lower offset
+                    if let Some(new_offset) = new_position {
+                        if let Some(offset) = entry.position {
+                            if offset < new_offset {
+                                *entry = new_entry;
+                            }
+                        } else {
+                            *entry = new_entry;
+                        }
+                    }
+                }
+            });
+
+            let mut removed_times = Vec::new();
+            for (time, (cap, map)) in to_send.iter_mut() {
+                if !input.frontier.less_equal(time) {
+                    let mut session = output.session(cap);
+                    removed_times.push(time.clone());
+                    for (key, data) in map.drain() {
+                        // decode key and value
+                        match key_decoder_state.decode_key(&key) {
+                            Ok(decoded_key) => {
+                                let decoded_value = if data.value.is_empty() {
+                                    Ok(None)
+                                } else if let Ok(value) = value_decoder_state.decode_upsert_value(
+                                    &data.value,
+                                    data.position,
+                                    data.upstream_time_millis,
+                                ) {
+                                    if let Some(value) = value {
+                                        // prepend key to row
+                                        row_packer.extend_by_row(&decoded_key);
+                                        row_packer.extend_by_row(&value);
+                                        Ok(Some(row_packer.finish_and_reuse()))
+                                    } else {
+                                        Ok(None)
+                                    }
+                                } else {
+                                    Err(())
```

Comment:
i think we could just log the error and continue here? which might simplify the rest of the logic as then you could treat the decoded value as a `Option<T>` instead of a `Resut<Option<T>>`?